core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -140,6 +200,362 @@ protected[actions] trait PrimitiveActions {
+
+    activationResponse
+      .map {
+        case Left(activationId) => // invocation timeout
+          session.logs += activationId
+          Left(ActivationResponse.whiskError(compositionActivationTimeout(activationId)))
+        case Right(activation) => // successful invocation
+          session.logs += activation.activationId
+          // activation.duration should be defined but this is not reflected by the type so be defensive
+          session.duration += activation.duration.getOrElse(activation.end.toEpochMilli - activation.start.toEpochMilli)
+          activation.annotations.get("limits").foreach { limitsAnnotation =>
+            limitsAnnotation.asJsObject.getFields("memory") match {
+              case Seq(JsNumber(memory)) =>
+                session.maxMemory = Math.max(session.maxMemory, memory.toInt)
+            }
+          }
+          Right(activation)
+      }
+      .recover { // invocation failure
+        case _ => Left(ActivationResponse.whiskError(compositionActivationFailure))
+      }
+  }
+
+  /**
+   * Creates an activation for a composition and writes it back to the datastore.
+   * Returns the activation.
+   */
+  private def completeActivation(user: Identity, session: Session, response: ActivationResponse)(
+    implicit transid: TransactionId): WhiskActivation = {
+
+    // compute max memory
+    val sequenceLimits = Parameters(
+      WhiskActivation.limitsAnnotation,
+      ActionLimits(session.action.limits.timeout, MemoryLimit(session.maxMemory MB), session.action.limits.logs).toJson)
+
+    // set causedBy if not topmost
+    val causedBy = session.cause.map { _ =>
+      Parameters(WhiskActivation.causedByAnnotation, JsString(Exec.SEQUENCE))
+    }
+
+    val end = Instant.now(Clock.systemUTC())
+
+    // create the whisk activation
+    val activation = WhiskActivation(
+      namespace = user.namespace.toPath,
+      name = session.action.name,
+      user.subject,
+      activationId = session.activationId,
+      start = session.start,
+      end = end,
+      cause = session.cause,
+      response = response,
+      logs = ActivationLogs(session.logs.map(_.asString).toVector),
+      version = session.action.version,
+      publish = false,
+      annotations = Parameters(WhiskActivation.topmostAnnotation, JsBoolean(session.cause.isEmpty)) ++
+        Parameters(WhiskActivation.pathAnnotation, JsString(session.action.fullyQualifiedName(false).asString)) ++
+        Parameters(WhiskActivation.kindAnnotation, JsString(Exec.SEQUENCE)) ++
+        Parameters(WhiskActivation.conductorAnnotation, JsBoolean(true)) ++
+        causedBy ++
+        sequenceLimits,
+      duration = Some(session.duration))
+
+    logging.debug(this, s"recording activation '${activation.activationId}'")
+    WhiskActivation.put(activationStore, activation)(transid, notifier = None) onComplete {
+      case Success(id) => logging.info(this, s"recorded activation")
```

Comment:
logging.debug; to follow in general log reduction.

Reply:
or just drop it entirely, it's logged in the dbstore anyway.

Reply:
I am confused. The code already invokes `logging.debug`. Should this change?

Reply:
line 549:
 +      case Success(id) => logging.info(this, s"recorded activation")

Reply:
OK I see it.